enhancements/installer/coarse-grained-exit-codes.md

```diff
@@ -0,0 +1,162 @@
+---
+title: coarse-grained-exit-codes
+authors:
+  - "@deadsk2"
+reviewers:
+  - "@staebler"
+approvers:
+  - "@pdillon"
+  - "@sdodson"
+api-approvers: # in case of new or modified APIs or API extensions (CRDs, aggregated apiservers, webhooks, finalizers)
+  - TBD
+creation-date: 2022-03-16
+last-updated: 2022-03-16
+tracking-link: # link to the tracking ticket (for example: Jira Feature or Epic ticket) that corresponds to this enhancement
+  - TBD
+see-also:
+replaces:
+superseded-by:
+---
+
+# Coarse Grained Exit Codes from the Installer
+
+## Summary
+
+`openshift-installer` callers will now get more granular exit codes.
+Callers of `openshift-installer` can expect non-zero exit codes *that may change over time* as
+the `openshift-installer` evolves over time.
+`openshift-installer` provides *no guarantee* that exit codes will not change as result of changes like
+greater granularity, lower granularity, re-organization, or other needs.
+`openshift-installer` *does guarantee* that a particular exit code will not be re-used to have a
+different meaning in the same y-level version.
+`openshift-installer` will make reasonable efforts to avoid re-using a particular exit code to have
+different meanings across y-level versions, but for sufficiently compelling reasons may do so.
+
+## Motivation
+
+Automated categorization of failures can assist in efforts like
+1. reliability determination of provision on various clouds
+2. directing effort to proper teams depending on failure modes
+3. tracking of which installation stages have problems on various platforms
+4. automated retry of infrastructure failures by CI step registry
+
+These are reasons why the data is useful and by providing it, other teams will be able to build these
+secondary system.
+Other teams *must properly handle previously unknown exit codes*.
+
+### Goals
+
+1. Coarse grained (order of 10 or less values), stage-based exit codes to assist in categorizing
+   failures, but not debugging failures.
+
+### Non-Goals
+
+1. Detailed, machine-parseable, cause-based failure information is outside the scope of this enhancement.
+
+## Proposal
+
+The installer knows what stage of the installation has failed when it exits.
+The stages include
+1. install-config handling
+2. infrastructure creation
+3. bootstrapping
+4. wait-for-cluster-install
+5. destroy
```

Comment:
What is destroy? Do you mean the step of destroying the bootstrap? Or do yo mean destroying the entire cluster?